File: summerpep/lec11.cpp
#include<bits/stdc++.h>
using namespace std;

class Node
{
public:
    int data;
    Node* left;
    Node* right;
    Node(int data){
        this->data = data;
        this->left = nullptr;
        this->right = nullptr;
    }
};

class binarytree{
    public:
    int ind = 0;

    Node* buildtree(vector<int> v){
        if(ind > v.size()){
            return nullptr;
        }
        if(v[ind] == -1){
            ind++;return nullptr;
        }
        Node* newnode = new Node(v[ind]);
        ind++;
        newnode->left = buildtree(v);
        newnode->right = buildtree(v);
        return newnode;
    }

    void po(Node* a){
        if(a == nullptr){
            return;
        }
        cout<<a->data<<" ";
        po(a->left);
        po(a->right);
    }
};



int main(){
    binarytree bt;
    vector<int> v = { 1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1 };
    Node* root = bt.buildtree(v);
    bt.po(root);
    return 0;
}

File: summerpep/lec11.java
import java.util.*;

public class lec11 {

    static class Node {
        int data;
        Node left;
        Node right;

        Node(int data) {
            this.data = data;
            this.left = null;
            this.right = null;
        }

    }

    static class binarytree {
        static int idx;

        public static Node buildtree(int nodes[]) {
            if (idx > nodes.length) {
                return null;
            }
            if (nodes[idx] == -1) {
                idx++;
                return null;
            }
            Node newnode = new Node(nodes[idx]);
            idx++;
            newnode.left = buildtree(nodes);
            newnode.right = buildtree(nodes);
            return newnode;
        }

        public static void po(Node a) {
            if (a == null) {
                return;
            }
            // list.add(a.val);
            System.out.print(a.data + " ");
            po(a.left);
            po(a.right);
        }

        // public static void lot(Node a) {
        //     if (a == null) {
        //         return;
        //     }
        //     Queue<Node> q = new LinkedList<>();
        //     q.add(a);
        //     q.add(null);

        //     while (!q.isEmpty()) {

        //         if (q.peek() == null) {
        //             System.out.println();

        //         }

        //         else{
        //             while (!q.isEmpty()) {
        //                 System.out.println(q.peek().data+" ");
        //                 if (q.peek().left != null) {
        //                         q.add(q.peek().left);
        //                     }
        //                     if (q.peek().right != null) {
        //                         q.add(q.peek().right);
        //                     }
        //                     q.remove();
        //                 }
        //             }
        //         }

        //         // else{
                
        //         // System.out.print(q.peek().data + " ");
        //         // if (q.peek().left != null) {
        //         //     q.add(q.peek().left);
        //         // }
        //         // if (q.peek().right != null) {
        //         //     q.add(q.peek().right);
        //         // }
        //         // q.remove();

        //         // }

        //     }
        // public void topv(Node root){
        //     stacks<Integer> s = new LinkedList<>();
        //     for(Node i = root;i!=null;i=i.left){

        //     }
        // }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        // int a = sc.nextInt();
        int[] nodes = { 1, 2, 4, -1, -1, 5, -1, -1, 3, -1, 6, -1, -1 };
        binarytree tree = new binarytree();
        Node root = tree.buildtree((nodes));
        // System.out.println(root);
        tree.po(root);
        // tree.lot(root);

    }
}
